File: js/commands/fs.js
/* ROOT OF CRIME – filesystem commands */

Object.assign(Terminal, {
  _cmdCd(args) {
    const target = args[0] || '~';
    const full = VFS.abs(this.cwd, target, this.home);
    const res = VFS.resolve(this.vfs, '/', full);
    if (!res) return this._err(`cd: ${target}: No such file or directory`);
    if (res.node.type !== 'dir') return this._err(`cd: ${target}: Not a directory`);
    if (this.ctx?.jail && !full.startsWith(this.ctx.jail) && full !== this.ctx.jail) {
      return this._err(`cd: permission denied: cannot leave ${this.ctx.jail} during this task`);
    }
    this.cwd = full;
    this._updatePrompt();
    return this._ok('');
  },

  _cmdLs(args) {
    const flags = new Set();
    const paths = [];
    args.forEach((a) => {
      if (a.startsWith('-') && a !== '-') {
        a.slice(1).split('').forEach((c) => flags.add(c));
      } else paths.push(a);
    });
    const long = flags.has('l');
    const all = flags.has('a') || flags.has('A');
    const targets = paths.length ? paths : ['.'];

    const blocks = [];
    for (const t of targets) {
      const res = VFS.resolve(this.vfs, this.cwd, t, this.home);
      if (!res) return this._err(`ls: cannot access '${t}': No such file or directory`);
      if (res.node.type === 'file') {
        blocks.push(long ? VFS.formatLong(VFS.basename(res.path), res.node) : VFS.basename(res.path));
        continue;
      }
      let names = Object.keys(res.node.children).sort();
      if (!all) names = names.filter((n) => !n.startsWith('.'));
      if (all) names = ['.', '..', ...names];
      if (long) {
        const lines = [`total ${names.length * 4}`];
        names.forEach((n) => {
          if (n === '.') lines.push(VFS.formatLong('.', res.node));
          else if (n === '..') lines.push('drwxr-xr-x 1 root       root          4096 Aug 14 09:00 ..');
          else lines.push(VFS.formatLong(n, res.node.children[n]));
        });
        blocks.push(lines.join('\n'));
      } else {
        blocks.push(names.filter((n) => n !== '.' && n !== '..').join('  '));
      }
    }
    return this._ok(blocks.join('\n\n'));
  },

  _cmdMkdir(args) {
    const paths = args.filter((a) => !a.startsWith('-'));
    if (!paths.length) return this._err('mkdir: missing operand');
    for (const a of paths) {
      const r = VFS.mkdir(this.vfs, this.cwd, a, this.home);
      if (!r.ok) return this._err(`mkdir: ${r.error}`);
    }
    return this._ok('');
  },

  _cmdRmdir(args) {
    if (!args.length) return this._err('rmdir: missing operand');
    for (const a of args) {
      const r = VFS.unlink(this.vfs, this.cwd, a, this.home);
      if (!r.ok) return this._err(`rmdir: ${r.error}`);
    }
    return this._ok('');
  },

  _cmdTouch(args) {
    if (!args.length) return this._err('touch: missing file operand');
    for (const a of args) {
      const r = VFS.touch(this.vfs, this.cwd, a, this.home);
      if (!r.ok) return this._err(`touch: ${r.error}`);
    }
    return this._ok('');
  },

  _cmdRm(args) {
    const recursive = args.some((a) => a === '-r' || a === '-rf' || a === '-fr');
    const force = args.some((a) => a === '-f' || a === '-rf' || a === '-fr');
    const paths = args.filter((a) => !a.startsWith('-'));
    if (!paths.length) return this._err('rm: missing operand');
    for (const p of paths) {
      const res = VFS.resolve(this.vfs, this.cwd, p, this.home);
      if (!res) {
        if (force) continue;
        return this._err(`rm: cannot remove '${p}': No such file or directory`);
      }
      if (res.path === '/' || res.path === '/home' || res.path === '/etc') {
        return this._err(`rm: refusing to remove '${p}'`);
      }
      const r = recursive
        ? VFS.rmRecursive(this.vfs, this.cwd, p, this.home)
        : VFS.unlink(this.vfs, this.cwd, p, this.home);
      if (!r.ok) return this._err(`rm: ${r.error}`);
      if (this.ctx) {
        this.ctx.removed = this.ctx.removed || [];
        this.ctx.removed.push(res.path);
      }
    }
    return this._ok('');
  },

  _cmdCp(args) {
    const paths = args.filter((a) => !a.startsWith('-'));
    if (paths.length < 2) return this._err('cp: missing file operand');
    const dest = paths.pop();
    const src = paths[0];
    const got = this._readFile(src);
    if (got.error) return this._err(got.error.replace(/^cat/, 'cp'));
    const destRes = VFS.resolve(this.vfs, this.cwd, dest, this.home);
    const destPath = destRes && destRes.node.type === 'dir'
      ? VFS.join(destRes.path, VFS.basename(VFS.abs(this.cwd, src, this.home)))
      : dest;
    const space = this._guestSpace(VFS.abs(this.cwd, destPath, this.home), (got.text || '').length);
    if (!space.ok) return this._err('cp: ' + space.error);
    const r = VFS.write(this.vfs, this.cwd, destPath, got.text, { home: this.home });
    if (!r.ok) return this._err(`cp: ${r.error}`);
    return this._ok('');
  },

  _cmdMv(args) {
    const paths = args.filter((a) => !a.startsWith('-'));
    if (paths.length < 2) return this._err('mv: missing file operand');
    const dest = paths[paths.length - 1];
    const srcs = paths.slice(0, -1);
    const destRes = VFS.resolve(this.vfs, this.cwd, dest, this.home);
    if (srcs.length > 1 && !(destRes && destRes.node.type === 'dir')) {
      return this._err(`mv: target '${dest}' is not a directory`);
    }
    for (const src of srcs) {
      const got = this._readFile(src);
      if (got.error) return this._err(got.error.replace(/^cat/, 'mv'));
      const destPath = destRes && destRes.node.type === 'dir'
        ? VFS.join(destRes.path, VFS.basename(VFS.abs(this.cwd, src, this.home)))
        : dest;
      const space = this._guestSpace(VFS.abs(this.cwd, destPath, this.home), (got.text || '').length);
      if (!space.ok) return this._err('mv: ' + space.error);
      const w = VFS.write(this.vfs, this.cwd, destPath, got.text, { home: this.home });
      if (!w.ok) return this._err(`mv: ${w.error}`);
      VFS.unlink(this.vfs, this.cwd, src, this.home);
    }
    return this._ok('');
  },

  _cmdChmod(args) {
    if (args.length < 2) {
      return this._err('chmod: missing operand\nTry: chmod 644 file   or   chmod 644 *');
    }
    const mode = args[0];
    const targets = args.slice(1);
    let count = 0;
    for (const t of targets) {
      const r = VFS.chmod(this.vfs, this.cwd, t, mode, this.home);
      if (!r.ok) return this._err(`chmod: ${r.error}`);
      count++;
    }
    if (this.ctx) this.ctx.chmodCount = (this.ctx.chmodCount || 0) + count;
    return this._ok(count > 1 ? `mode of ${count} files changed to ${mode}` : `mode of '${targets[0]}' changed to ${mode}`);
  },

  _cmdFind(args) {
    let start = '.';
    let name = null;
    let type = null;
    for (let i = 0; i < args.length; i++) {
      if (args[i] === '-name' && args[i + 1]) {
        name = args[i + 1].replace(/^['"]|['"]$/g, '');
        i++;
      } else if (args[i] === '-type' && args[i + 1]) {
        type = args[i + 1] === 'd' ? 'dir' : 'file';
        i++;
      } else if (!args[i].startsWith('-')) {
        start = args[i];
      }
    }
    const startRes = VFS.resolve(this.vfs, this.cwd, start, this.home);
    if (!startRes) return this._err(`find: '${start}': No such file or directory`);
    const nameRe = name ? (() => {
      const re = name.replace(/\./g, '\\.').replace(/\*/g, '.*').replace(/\?/g, '.');
      return new RegExp('^' + re + '$');
    })() : null;
    const hits = [];
    VFS.walk(startRes.node, startRes.path, (node, path) => {
      if (type && node.type !== type) return;
      if (nameRe && !nameRe.test(VFS.basename(path))) return;
      hits.push(path);
    });
    if (this.ctx) this.ctx.usedFind = true;
    return this._ok(hits.join('\n'));
  },

  _cmdFile(args) {
    if (!args.length) return this._err('file: missing operand');
    const lines = args.map((a) => {
      const res = VFS.resolve(this.vfs, this.cwd, a, this.home);
      if (!res) return `${a}: cannot open (No such file or directory)`;
      if (res.node.type === 'dir') return `${a}: directory`;
      const c = res.node.content;
      if (c.startsWith('ELF')) return `${a}: ELF 64-bit LSB executable`;
      if (c.startsWith('[binary') || c.includes('JPEG') || a.endsWith('.jpg')) return `${a}: JPEG image data`;
      return `${a}: ASCII text`;
    });
    return this._ok(lines.join('\n'));
  }
});

Terminal.define('cd', function (args) { return this._cmdCd(args); });
Terminal.define('ls', function (args) { return this._cmdLs(args); });
Terminal.define('mkdir', function (args) { return this._cmdMkdir(args); });
Terminal.define('rmdir', function (args) { return this._cmdRmdir(args); });
Terminal.define('touch', function (args) { return this._cmdTouch(args); });
Terminal.define('rm', function (args) { return this._cmdRm(args); });
Terminal.define('cp', function (args) { return this._cmdCp(args); });
Terminal.define('mv', function (args) { return this._cmdMv(args); });
Terminal.define('chmod', function (args) { return this._cmdChmod(args); });
Terminal.define('find', function (args) { return this._cmdFind(args); });
Terminal.define('file', function (args) { return this._cmdFile(args); });

File: crates/roc-kernel/src/shell/commands/fs.rs
//! Filesystem commands. Matches `js/commands/fs.js`.

use regex::Regex;

use crate::vfs::{self, WriteOpts};

use super::super::exec::Shell;
use super::super::output::Output;

impl Shell {
    pub(crate) fn cmd_cd(&mut self, args: &[String]) -> Output {
        let target = args.first().map(String::as_str).unwrap_or("~");
        let full = vfs::abs(&self.cwd, target, &self.home);
        let Some(res) = vfs::resolve(&self.vfs, "/", &full, &self.home) else {
            return Output::err(format!("cd: {target}: No such file or directory"));
        };
        if !res.node.is_dir() {
            return Output::err(format!("cd: {target}: Not a directory"));
        }
        if let Some(jail) = &self.ctx.jail {
            if !full.starts_with(jail) && full != *jail {
                return Output::err(format!(
                    "cd: permission denied: cannot leave {jail} during this task"
                ));
            }
        }
        self.cwd = full;
        Output::ok("")
    }

    pub(crate) fn cmd_ls(&mut self, args: &[String]) -> Output {
        let mut flags = std::collections::HashSet::new();
        let mut paths = Vec::new();
        for a in args {
            if a.starts_with('-') && a != "-" {
                for c in a.chars().skip(1) {
                    flags.insert(c);
                }
            } else {
                paths.push(a.as_str());
            }
        }
        let long = flags.contains(&'l');
        let all = flags.contains(&'a') || flags.contains(&'A');
        let default = ["."];
        let targets: Vec<&str> = if paths.is_empty() {
            default.to_vec()
        } else {
            paths
        };

        let mut blocks = Vec::new();
        for t in targets {
            let Some(res) = vfs::resolve(&self.vfs, &self.cwd, t, &self.home) else {
                return Output::err(format!("ls: cannot access '{t}': No such file or directory"));
            };
            if res.node.is_file() {
                let name = vfs::basename(&res.path);
                blocks.push(if long {
                    vfs::format_long(&name, res.node)
                } else {
                    name
                });
                continue;
            }
            let mut names: Vec<String> = res
                .node
                .children()
                .map(|c| c.keys().cloned().collect())
                .unwrap_or_default();
            names.sort();
            if !all {
                names.retain(|n| !n.starts_with('.'));
            }
            if all {
                let mut with_dots = vec![".".to_string(), "..".to_string()];
                with_dots.extend(names);
                names = with_dots;
            }
            if long {
                let mut lines = vec![format!("total {}", names.len() * 4)];
                for n in &names {
                    if n == "." {
                        lines.push(vfs::format_long(".", res.node));
                    } else if n == ".." {
                        lines.push(
                            "drwxr-xr-x 1 root       root          4096 Aug 14 09:00 .."
                                .to_string(),
                        );
                    } else {
                        let child = res.node.children().and_then(|c| c.get(n)).unwrap();
                        lines.push(vfs::format_long(n, child));
                    }
                }
                blocks.push(lines.join("\n"));
            } else {
                let shown: Vec<&str> = names
                    .iter()
                    .map(String::as_str)
                    .filter(|n| *n != "." && *n != "..")
                    .collect();
                blocks.push(shown.join("  "));
            }
        }
        Output::ok(blocks.join("\n\n"))
    }

    pub(crate) fn cmd_mkdir(&mut self, args: &[String]) -> Output {
        let paths: Vec<&str> = args
            .iter()
            .map(String::as_str)
            .filter(|a| !a.starts_with('-'))
            .collect();
        if paths.is_empty() {
            return Output::err("mkdir: missing operand");
        }
        for a in paths {
            if let Err(e) = vfs::mkdir(&mut self.vfs, &self.cwd, a, &self.home) {
                return Output::err(format!("mkdir: {}", e.message));
            }
        }
        Output::ok("")
    }

    pub(crate) fn cmd_rmdir(&mut self, args: &[String]) -> Output {
        if args.is_empty() {
            return Output::err("rmdir: missing operand");
        }
        for a in args {
            if let Err(e) = vfs::unlink(&mut self.vfs, &self.cwd, a, &self.home) {
                return Output::err(format!("rmdir: {}", e.message));
            }
        }
        Output::ok("")
    }

    pub(crate) fn cmd_touch(&mut self, args: &[String]) -> Output {
        if args.is_empty() {
            return Output::err("touch: missing file operand");
        }
        for a in args {
            if let Err(e) = vfs::touch(&mut self.vfs, &self.cwd, a, &self.home) {
                return Output::err(format!("touch: {}", e.message));
            }
        }
        Output::ok("")
    }

    pub(crate) fn cmd_rm(&mut self, args: &[String]) -> Output {
        let recursive = args.iter().any(|a| a == "-r" || a == "-rf" || a == "-fr");
        let force = args.iter().any(|a| a == "-f" || a == "-rf" || a == "-fr");
        let paths: Vec<&str> = args
            .iter()
            .map(String::as_str)
            .filter(|a| !a.starts_with('-'))
            .collect();
        if paths.is_empty() {
            return Output::err("rm: missing operand");
        }
        for p in paths {
            let res = vfs::resolve(&self.vfs, &self.cwd, p, &self.home);
            if res.is_none() {
                if force {
                    continue;
                }
                return Output::err(format!("rm: cannot remove '{p}': No such file or directory"));
            }
            let path = res.unwrap().path.clone();
            if path == "/" || path == "/home" || path == "/etc" {
                return Output::err(format!("rm: refusing to remove '{p}'"));
            }
            let result = if recursive {
                vfs::rm_recursive(&mut self.vfs, &self.cwd, p, &self.home)
            } else {
                vfs::unlink(&mut self.vfs, &self.cwd, p, &self.home)
            };
            if let Err(e) = result {
                return Output::err(format!("rm: {}", e.message));
            }
            self.ctx.removed.push(path);
        }
        Output::ok("")
    }

    pub(crate) fn cmd_cp(&mut self, args: &[String]) -> Output {
        let mut paths: Vec<&str> = args
            .iter()
            .map(String::as_str)
            .filter(|a| !a.starts_with('-'))
            .collect();
        if paths.len() < 2 {
            return Output::err("cp: missing file operand");
        }
        let dest = paths.pop().unwrap();
        let src = paths[0];
        let text = match self.read_file(src) {
            Ok(f) => f.text,
            Err(e) => return Output::err(super::super::exec::retarget_cat(&e, "cp")),
        };
        let dest_path = match vfs::resolve(&self.vfs, &self.cwd, dest, &self.home) {
            Some(res) if res.node.is_dir() => vfs::join(
                &res.path,
                &vfs::basename(&vfs::abs(&self.cwd, src, &self.home)),
            ),
            _ => dest.to_string(),
        };
        let abs_dest = vfs::abs(&self.cwd, &dest_path, &self.home);
        if let Err(e) = self.guest_space(&abs_dest, crate::vfs::js_len(&text) as u64) {
            return Output::err(format!("cp: {e}"));
        }
        if let Err(e) = vfs::write(
            &mut self.vfs,
            &self.cwd,
            &dest_path,
            &text,
            WriteOpts {
                home: Some(self.home.clone()),
                ..WriteOpts::default()
            },
        ) {
            return Output::err(format!("cp: {}", e.message));
        }
        Output::ok("")
    }

    pub(crate) fn cmd_mv(&mut self, args: &[String]) -> Output {
        let paths: Vec<&str> = args
            .iter()
            .map(String::as_str)
            .filter(|a| !a.starts_with('-'))
            .collect();
        if paths.len() < 2 {
            return Output::err("mv: missing file operand");
        }
        let dest = paths[paths.len() - 1];
        let srcs = &paths[..paths.len() - 1];
        let (dest_is_dir, dest_dir_path) = match vfs::resolve(&self.vfs, &self.cwd, dest, &self.home)
        {
            Some(r) if r.node.is_dir() => (true, Some(r.path.clone())),
            _ => (false, None),
        };
        if srcs.len() > 1 && !dest_is_dir {
            return Output::err(format!("mv: target '{dest}' is not a directory"));
        }
        for src in srcs {
            let text = match self.read_file(src) {
                Ok(f) => f.text,
                Err(e) => return Output::err(super::super::exec::retarget_cat(&e, "mv")),
            };
            let dest_path = if dest_is_dir {
                vfs::join(
                    dest_dir_path.as_ref().unwrap(),
                    &vfs::basename(&vfs::abs(&self.cwd, src, &self.home)),
                )
            } else {
                dest.to_string()
            };
            let abs_dest = vfs::abs(&self.cwd, &dest_path, &self.home);
            if let Err(e) = self.guest_space(&abs_dest, crate::vfs::js_len(&text) as u64) {
                return Output::err(format!("mv: {e}"));
            }
            if let Err(e) = vfs::write(
                &mut self.vfs,
                &self.cwd,
                &dest_path,
                &text,
                WriteOpts {
                    home: Some(self.home.clone()),
                    ..WriteOpts::default()
                },
            ) {
                return Output::err(format!("mv: {}", e.message));
            }
            let _ = vfs::unlink(&mut self.vfs, &self.cwd, src, &self.home);
        }
        Output::ok("")
    }

    pub(crate) fn cmd_chmod(&mut self, args: &[String]) -> Output {
        if args.len() < 2 {
            return Output::err(
                "chmod: missing operand\nTry: chmod 644 file   or   chmod 644 *",
            );
        }
        let mode = &args[0];
        let targets = &args[1..];
        let mut count = 0;
        for t in targets {
            if let Err(e) = vfs::chmod(&mut self.vfs, &self.cwd, t, mode, &self.home) {
                return Output::err(format!("chmod: {}", e.message));
            }
            count += 1;
        }
        self.ctx.chmod_count += count;
        if count > 1 {
            Output::ok(format!("mode of {count} files changed to {mode}"))
        } else {
            Output::ok(format!("mode of '{}' changed to {mode}", targets[0]))
        }
    }

    pub(crate) fn cmd_find(&mut self, args: &[String]) -> Output {
        let mut start = ".";
        let mut name: Option<&str> = None;
        let mut kind: Option<&str> = None;
        let mut i = 0;
        while i < args.len() {
            if args[i] == "-name" && i + 1 < args.len() {
                name = Some(args[i + 1].as_str());
                i += 2;
            } else if args[i] == "-type" && i + 1 < args.len() {
                kind = Some(if args[i + 1] == "d" { "dir" } else { "file" });
                i += 2;
            } else if !args[i].starts_with('-') {
                start = args[i].as_str();
                i += 1;
            } else {
                i += 1;
            }
        }
        let Some(start_res) = vfs::resolve(&self.vfs, &self.cwd, start, &self.home) else {
            return Output::err(format!("find: '{start}': No such file or directory"));
        };
        let name_re = if let Some(pat) = name {
            let pat = pat.trim_matches(|c| c == '\'' || c == '"');
            let escaped = pat.replace('.', r"\.").replace('*', ".*").replace('?', ".");
            match Regex::new(&format!("^{escaped}$")) {
                Ok(re) => Some(re),
                Err(e) => return Output::err(e.to_string()),
            }
        } else {
            None
        };
        let mut hits = Vec::new();
        vfs::walk(start_res.node, &start_res.path, |node, path| {
            if let Some(k) = kind {
                if node.kind() != k {
                    return;
                }
            }
            if let Some(re) = &name_re {
                if !re.is_match(&vfs::basename(path)) {
                    return;
                }
            }
            hits.push(path.to_string());
        });
        self.ctx.used_find = true;
        Output::ok(hits.join("\n"))
    }

    pub(crate) fn cmd_file(&mut self, args: &[String]) -> Output {
        if args.is_empty() {
            return Output::err("file: missing operand");
        }
        let lines: Vec<String> = args
            .iter()
            .map(|a| {
                let Some(res) = vfs::resolve(&self.vfs, &self.cwd, a, &self.home) else {
                    return format!("{a}: cannot open (No such file or directory)");
                };
                if res.node.is_dir() {
                    return format!("{a}: directory");
                }
                let c = res.node.content().unwrap_or("");
                if c.starts_with("ELF") {
                    format!("{a}: ELF 64-bit LSB executable")
                } else if c.starts_with("[binary") || c.contains("JPEG") || a.ends_with(".jpg") {
                    format!("{a}: JPEG image data")
                } else {
                    format!("{a}: ASCII text")
                }
            })
            .collect();
        Output::ok(lines.join("\n"))
    }
}
